src/atmega2560p/hal/port.rs

```diff
@@ -0,0 +1,185 @@
+//     RustDuino : A generic HAL implementation for Arduino Boards in Rust
+//     Copyright (C) 2021  Tulika Shukla, Indian Institute of Technology Kanpur
+//
+//     This program is free software: you can redistribute it and/or modify
+//     it under the terms of the GNU Affero General Public License as published
+//     by the Free Software Foundation, either version 3 of the License, or
+//     (at your option) any later version.
+//
+//     This program is distributed in the hope that it will be useful,
+//     but WITHOUT ANY WARRANTY; without even the implied warranty of
+//     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//     GNU Affero General Public License for more details.
+//
+//     You should have received a copy of the GNU Affero General Public License
+//     along with this program.  If not, see <https://www.gnu.org/licenses/>
+
+//! Various pins and ports in the ATMEGA2560P chip is controlled here.
+//! Section 13.2 to 13.4 of ATMEGA2560P datasheet.
+//! https://ww1.microchip.com/downloads/en/devicedoc/atmel-2549-8-bit-avr-microcontroller-atmega640-1280-1281-2560-2561_datasheet.pdf
+
+/// Core Crate functions required in the code for reading and writing to registers.
+use core::{
+    ptr::{read_volatile, write_volatile},
+    usize,
+};
+
+/// Represents the name of the ports in ATMEGA2560P , can vary from A-L leaving I.
+#[derive(Clone, Copy)]
+pub enum PortName {
+    A,
+    B,
+    C,
+    D,
+    E,
+    F,
+    G,
+    H,
+    J,
+    K,
+    L,
+}
+
+/// Each Port pin consists of 3 registers : PINx, DDRx, PORTx .
+/// These registors control the ports.
+/// ```DDRx:  Data direction register```
+///     The DDRx Register controls the direction of a particular pin.
+/// ```PORTx: Data register```
+///     PORTxn is written logic one when the pin is configured as an input pin.
+///     PORTxn is written logic zero when the pin is configured as an output pin.
+/// ```PINx:  Port input pins```
+///     Writing a logic one to PINxn toggles the value of PORTxn, independent on the value of DDRxn.
+///     This will control the I/O state of pins.    
+pub struct Port {
+    pin: u8,
+    ddr: u8,
+    port: u8,
+}
+
+/// The structure Pin contains the address of the port to which the pin belongs and the pin's number.
+pub struct Pin {
+    port: *mut Port,
+    pin: usize,
+}
+
+/// Type ```IOMode```
```

Comment:
Why triple ticks?

Reply:
I was correlating it with MARKDOWN. Sorry for that, adjusted everywhere.